### PR TITLE
web/tracking: extract user and session tracking from EventLogger

### PR DESCRIPTION
**Pull request by bobheadxi**
`EventLogger` is currently a very overloaded mechanism that handles all sorts of tracking mechanisms and cookie configurations on top of event logging.

This change extracts user tracking (anonymous user ID, cohorts, and device ID) and session tracking (URL stuff) from `EventLogger` into separate, isolated classes. Functionally everything should be the same, but streamlined significantly. Most importantly, this unblocks us so that we can use these mechanisms for introducing a client implementation for the [new telemetry framework](https://docs.sourcegraph.com/dev/background-information/telemetry) while preserving existing behaviour.

This only changes `client/web` - it seems this is copy-pasted into several places, we can work on centralizing them all later.

Part of https://github.com/sourcegraph/sourcegraph/issues/56920
Enables https://github.com/sourcegraph/sourcegraph/pull/57939

## Test plan

`sg start dotcom`, then:

```
select * from event_logs where source = 'WEB' order by timestamp desc limit 3;
```

### CODE REVIEW

File: client/web/src/tracking/SessionTracker.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    public getOriginalReferrer(): string {
+        if (!this.isSourcegraphDotComMode) {
+            return ''
+        }
+        /**
+         * Gets the original referrer from the cookie or, if it doesn't exist, the
+         * mkto_referrer from the URL.
+         */
+        this.originalReferrer =
+            this.originalReferrer ||
```
Comment (bobheadxi) [inline client/web/src/tracking/SessionTracker.ts:49]: These getters all do this - might seem a bit strange, but matches the existing behaviour of checking the cookie again if the value is not set

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #57884** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57884" target="_blank">Graphite</a>  👈
    * **PR #57939** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57939" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57884?utm_source=stack-comment).

File: client/web/src/tracking/SessionTracker.ts (hunk 1 above)
Comment (nathan-downs) [reply client/web/src/tracking/SessionTracker.ts]: Yep, I believe this was to ensure that we're getting the latest information from the cookie.

Comment (bobheadxi) [reply client/web/src/tracking/SessionTracker.ts]: The way this was written means that we only get the latest information _if_ we've never managed to get a value from the cookie before - if we got a value before, no subsequent call will check the cookie again. Is that what you intend here?

Comment (nathan-downs) [reply client/web/src/tracking/SessionTracker.ts]: Sorry, my response before wasn't clear/correct in this context (I was thinking about the refresh functionality).

As written this is the correct functionality for what is intended.

File: client/web/src/tracking/userTracker.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        // Use cookies instead of localStorage so that the ID can be shared with subdomains (about.sourcegraph.com).
+        // Always set to renew expiry and migrate from localStorage
+        cookies.set(ANONYMOUS_USER_ID_KEY, anonymousUserID, userCookieSettings)
+        localStorage.removeItem(ANONYMOUS_USER_ID_KEY)
+
+        if (cohortID) {
+            cookies.set(COHORT_ID_KEY, cohortID, userCookieSettings)
+        }
+
+        let deviceID = cookies.get(DEVICE_ID_KEY)
```
Comment (nathan-downs) [inline client/web/src/tracking/userTracker.ts:68]: I don't think we ever set this anywhere else ever, so practically this is only ever the anonymous_user_id.

**Review by nathan-downs — APPROVED** [review]
The logically extraction of functionality looks good to me, just a pending question about the utility/functionality of device_id and device_session_id, but that can be solved in a cleanup PR.

File: client/web/src/tracking/userTracker.ts (hunk 2 above)
Comment (bobheadxi) [reply client/web/src/tracking/userTracker.ts:68]: I'll follow up to remove these two entirely

Comment (bobheadxi) [reply client/web/src/tracking/userTracker.ts:68]: actually, note that the "device session ID" is in the new telemetry gateway schema:

```
  // Device session ID to identify the user's session.
  optional string device_session_id = 6;
```

https://docs.google.com/document/d/1F6yFUVX_DZ_tT76NpGJL6_jlm_THQyVikRRRjNDNsyg/edit#bookmark=id.ywv5ix2bu6i2

Should we remove this?

**Comment by bobheadxi** [comment]
manual backport: https://github.com/sourcegraph/sourcegraph/pull/58029

**Pull request merged**